Added init db scripts, simple model, route for tags

MaxSukhovetskyi opened the pull request:
Added init db scripts, simple model, route for tags

MaxSukhovetskyi committed:
Created tag model, add arguments for db drop, create empty tables to inialize.py, updated README.md

MaxSukhovetskyi committed:
Created basic GET route, controler for tags

MaxSukhovetskyi committed:
Created custom base class for shared functionality

MaxSukhovetskyi committed:
Removed MyModel model and its controler

MaxSukhovetskyi committed:
Removed MyModel model and its controler

MaxSukhovetskyi committed:
Merge branch 'db_init_tags_model' of https://github.com/Social-projects-Rivne/Rv-041WebUI-python into db_init_tags_model

MaxSukhovetskyi committed:
change user password to 12345678, restrict acsces to route tag(only GET)

MaxSukhovetskyi committed:
Not finished test data insert script

MaxSukhovetskyi committed:
fixed create user syntax error

MaxSukhovetskyi committed:
Created test database ORM structure, insert exemple data script
Added --reset attribute to initialize_db script for droping and creating database
Created test models with relations
Added library Faker to generate random example data(locked seed)
Created scrip to fill test models using Faker`s data
Created many to many relationship between Tag and Restaurant tables

EvanAntoni committed:
fix tag association table name

Commit:
fix tag association table name

MaxSukhovetskyi committed:
fix to allow CORS for react

MaxSukhovetskyi committed:
Merge branch 'test_data' of https://github.com/Social-projects-Rivne/Rv-041WebUI-python into test_data

MaxSukhovetskyi committed:
Added routes, controlers for restaurant, menu

MaxSukhovetskyi committed:
Added unique id for use in react key section

MaxSukhovetskyi committed:
Change get restaurant by id url to /restaurant/{id}, removed menu_id field from output json

MaxSukhovetskyi committed:
added block for non integer id call for restaurant

MaxSukhovetskyi committed:
--reset option help added to readme

serge30 commented on README.md:
```diff
@@ -47,6 +47,11 @@ and then restart postqresql service
 
 `(venv) initialize_easyrest_db development.ini`
```
I'd add `--fill` argument here, as fro development environment it's natural to have populated DB.

MaxSukhovetskyi replied:
I already implement this in other branch so it will be in next pull request.

serge30 replied:
My initial comment was to update the command in documentation as:
```
(venv) initialize_easyrest_db --fill development.ini
```

MaxSukhovetskyi replied:
Added

serge30 commented on README.md:
```diff
@@ -98,15 +103,14 @@ alias envon=`source ~/venv/<Your venv name>/bin/activate`
 10. Initialize db
 
-`(%VENV%) initialize_easyrest_db development.ini`
+`(venv) initialize_easyrest_db development.ini`
```
The same as above.

MaxSukhovetskyi replied:
Same as above, i also added --reset for db drop and create empty db

serge30 commented on easyrest/models/meta.py:
```diff
@@ -1,4 +1,4 @@
-from sqlalchemy.ext.declarative import declarative_base
```
Please do not forget about docstrings fro modules.

serge30 commented on easyrest/models/meta.py:
```diff
@@ -13,4 +13,16 @@
+
+
+@as_declarative(metadata=metadata)
+class Base(object):
```
Docstring fro the class.

serge30 commented on easyrest/models/tag_model.py:
```diff
@@ -8,11 +8,7 @@
-class MyModel(Base):
-    __tablename__ = 'models'
+class Tag(Base):
+    __tablename__ = 'Tag'
```
It's common to use  [snake case](https://en.wikipedia.org/wiki/Snake_case) for table names.

MaxSukhovetskyi replied:
I`ve rewrite tag model acording to snacke case.

serge30 commented on easyrest/scripts/initialize_db.py:
```diff
@@ -8,12 +8,20 @@
-    model = models.mymodel.MyModel(name='one', value=1)
-    dbsession.add(model)
+
+def fill_models(dbsession):
```
Docsrtrings

serge30 commented on easyrest/scripts/initialize_db.py:
```diff
@@ -8,12 +8,20 @@
+        dbsession.add(model_item)
+
+
+def drop_models(dbsession):
```
Doc strings.

serge30 commented on easyrest/scripts/initialize_db.py:
```diff
@@ -33,16 +51,14 @@ def main(argv=sys.argv):
-            setup_models(dbsession)
+            if args.drop:
+                drop_models(dbsession)
+                print('Droped')
```
We are using Python 2.x where `print` is not function, but statement. Parentheses are redundant  here. Also make the message more verbose (e.g. "Database has been dropped")

serge30 commented on easyrest/scripts/initialize_db.py:
```diff
@@ -33,16 +51,14 @@ def main(argv=sys.argv):
+                print('Droped')
+            else:
+                setup_models(dbsession)
+                print('Created')
```
The same as above.

serge30 commented on easyrest/scripts/initialize_db.py:
```diff
@@ -33,16 +51,14 @@ def main(argv=sys.argv):
+                print('Created')
+            if args.fill:
+                fill_models(dbsession)
+                print("filled")
```
The same as above. And message might be something like: "Database has been populated by testing data"

serge30 commented on easyrest/views/tag_controler.py:
```diff
@@ -0,0 +1,21 @@
+from pyramid.view import view_config
```
Docstrings.

serge30 commented on easyrest/views/tag_controler.py:
```diff
@@ -0,0 +1,21 @@
+            raise HTTPNotFound()
+        dict_objs = [tag.as_dict() for tag in tags]
+    except DBAPIError:
+        return Response(db_err_msg, content_type='text/plain', status=500)
```
I do not see where `db_err_msg` variable is defined. Anyway returning plain text with error message is not very good idea, as this output will be handled by JS on frontend side so make it more machine readable would be great. I'd propose for all JSON use the next format:
```
{
  "data": ... // result if no error
  "success": true/false // is request was succeeded
  "error": ... // error message if any
}
```

serge30 replied:
It's something was not fixed from last review

MaxSukhovetskyi replied:
I`ve remove try, except block entirely.

serge30 reviewed (changes requested):
Docstrings, docstrings, docstrings... for modules, classes, methods and functions

RazSKg committed:
added docstrings

MaxSukhovetskyi committed:
Change prints acording to python 2.7

MaxSukhovetskyi committed:
Added more informative messages

MaxSukhovetskyi committed:
Merged with routes_for_test_data branch

MaxSukhovetskyi committed:
Merge branch 'db_init_tags_model' of https://github.com/Social-projects-Rivne/Rv-041WebUI-python into db_init_tags_model

MaxSukhovetskyi committed:
Make up to date with remote

MaxSukhovetskyi committed:
Moved CORS fix to __init__.py

MaxSukhovetskyi committed:
Added more docstrings, make docstrings more informative

MaxSukhovetskyi committed:
Add wrap function for making json response

MaxSukhovetskyi committed:
Change tag controler to output json in new style, added tag controller docstring, change json formation to common function wrap

serge30 commented on easyrest/__init__.py:
```diff
@@ -1,4 +1,18 @@
 from pyramid.config import Configurator
```
Docstring

serge30 commented on easyrest/models/Menu_item.py:
```diff
@@ -0,0 +1,28 @@
+from .meta import Base
+
+
+class Menu_item(Base):
```
All models names should follow PEP8 rules of class naming.

serge30 commented on easyrest/models/Menu_item.py:
```diff
@@ -0,0 +1,28 @@
+"""
```
https://www.python.org/dev/peps/pep-0008/#package-and-module-names
>Modules should have short, all-lowercase names. Underscores can be used in the module name if it improves readability. Python packages should also have short, all-lowercase names, although the use of underscores is discouraged.

serge30 commented on easyrest/models/Menu.py:
```diff
@@ -0,0 +1,26 @@
+    id = Column(Integer, primary_key=True)
+    name = Column(Text)
+
+    # restaurant = relationship("Restaurant")
```
Why it's commented?

serge30 commented on easyrest/models/Tag_association.py:
```diff
@@ -0,0 +1,27 @@
+"""
```
PEP8: module name

serge30 commented on easyrest/models/Tag_association.py:
```diff
@@ -0,0 +1,27 @@
+from .meta import Base
+
+
+class Tag_association(Base):
```
PEP8: Class name

serge30 commented on easyrest/scripts/initialize_db.py:
```diff
@@ -1,48 +1,93 @@
+                fill_models(dbsession)
+                print 'Database has been populated by testing data'
+    except OperationalError as e:
+        print 'OperationalError: {}' % (e)
```
It's incorrect syntax for string formatting:
```
In [1]: e = 'test'

In [2]: print 'OperationalError: {}' % (e)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-bee065ca4472> in <module>()
----> 1 print 'OperationalError: {}' % (e)

TypeError: not all arguments converted during string formatting
```

serge30 commented on easyrest/scripts/insert_exemple_data.py:
```diff
@@ -0,0 +1,64 @@
+"""
```
Typo int the module's name: `...example...`

serge30 commented on easyrest/scripts/insert_exemple_data.py:
```diff
@@ -0,0 +1,64 @@
+"""
+
+from ..models import *
+from random import randint, seed
```
PEP8: https://www.python.org/dev/peps/pep-0008/#imports

serge30 commented on easyrest/scripts/insert_exemple_data.py:
```diff
@@ -0,0 +1,64 @@
+    Tags_models = [Tag(**tag) for tag in Tags]
+
+    for i in range(10):
+        rest = {
```
I'm not sure it's a great idea to use fake data. Probably it's suitable for unit tests, but not for  dev DB.

I'd rather use hard-coded values, as it is something we may control, and it will look much better on demos.

serge30 commented on easyrest/views/menu_controler.py:
```diff
@@ -0,0 +1,34 @@
+
+from sqlalchemy.exc import DBAPIError
+
+from json import dumps
```
PEP8: imports order

serge30 commented on easyrest/views/menu_controler.py:
```diff
@@ -0,0 +1,34 @@
+
+def asign_items(menu):
+    menu_dict = menu.as_dict()
+    menu_dict["id"] = "menuId" + str(menu_dict["id"])
```
Please use string formatting:
```
menu_dict["id"] = "menuId%s" % menu_dict["id"]
```

serge30 commented on easyrest/views/menu_controler.py:
```diff
@@ -0,0 +1,34 @@
+@view_config(route_name='get_menu', renderer='json', request_method='GET')
+def get_menu_controler(request):
+    rest_id = request.matchdict['id']
+    rest = request.dbsession.query(Restaurant).filter(Restaurant.id == rest_id).all()
```
SQLAlchemy has useful method to get row by id: [Query.get](https://docs.sqlalchemy.org/en/latest/orm/query.html#sqlalchemy.orm.query.Query.get)

serge30 commented on easyrest/views/menu_controler.py:
```diff
@@ -0,0 +1,34 @@
+def get_menu_controler(request):
+    rest_id = request.matchdict['id']
+    rest = request.dbsession.query(Restaurant).filter(Restaurant.id == rest_id).all()
+    if len(rest) == 0:
```
To check whether list/tuple/dict/set is empty in Python indirect conversion to `bool` is used:
```
if not rest:
    ...
```

serge30 commented on easyrest/views/menu_controler.py:
```diff
@@ -0,0 +1,34 @@
+    if len(rest) == 0:
+        return Response(body=dumps({"data": [], "name": "get_menu" }))
+    menu_dict = asign_items(rest[0].menu)
+    response = Response(body=dumps({"data":menu_dict, "name": "get_menu" }))
```
You already have render assigned to this view, returning just `dict` is enough. Pyramid will convert it to the JSON automatically as `renderer='json'` is set.

serge30 commented on easyrest/views/restaurant_controler.py:
```diff
@@ -0,0 +1,49 @@
+from pyramid.view import view_config
```
Docstring.

serge30 commented on easyrest/views/restaurant_controler.py:
```diff
@@ -0,0 +1,49 @@
+
+from sqlalchemy.exc import DBAPIError
+
+from json import dumps
```
PEP8: Import order

serge30 commented on easyrest/views/restaurant_controler.py:
```diff
@@ -0,0 +1,49 @@
+
+def asign_tags(rests):
+    rests_dict = [rest.as_dict() for rest in rests]
+    for i, rest in enumerate(rests):
```
You don't need `enumerate` dict is mutable type so it's returned by pointer and can be updated in-place.
```
for restaurant in restaurants:
    del restaurant["menu_id"]
    restaurant["id"] = ...
```

serge30 commented on easyrest/views/restaurant_controler.py:
```diff
@@ -0,0 +1,49 @@
+def asign_tags(rests):
+    rests_dict = [rest.as_dict() for rest in rests]
+    for i, rest in enumerate(rests):
+        del(rests_dict[i]["menu_id"])
```
`del` is not function, but statement. Parentheses are not needed here.

serge30 commented on easyrest/views/restaurant_controler.py:
```diff
@@ -0,0 +1,49 @@
+        tags = rest.tag
+        tags_list = [tag.as_dict() for tag in tags]
+        for tag in tags_list:
+            tag["id"] = "tagId" + str(tag["id"])
```
Why do you update ids here?

MaxSukhovetskyi replied:
To use as unique key in React components.

serge30 commented on easyrest/views/restaurant_controler.py:
```diff
@@ -0,0 +1,49 @@
+
+@view_config(route_name='get_all_restaurants', renderer='json', request_method='GET')
+def get_all_restaurant_controler(request):
+    def return_all():
```
Why do you need nested function here?

serge30 commented on easyrest/views/restaurant_controler.py:
```diff
@@ -0,0 +1,49 @@
+        rests = request.dbsession.query(Restaurant).all()
+        rests_dict = asign_tags(rests)
+        
+        response = Response(body=dumps({"data":rests_dict, "name": "get_restaurant" }))
```
Return dict, Pyramid will do the rest.

MaxSukhovetskyi replied:
Change to pyramid return, and CORS fix works fine

serge30 commented on easyrest/views/restaurant_controler.py:
```diff
@@ -0,0 +1,49 @@
+
+@view_config(route_name='get_restaurant', renderer='json', request_method='GET')
+def get_restaurant_controler(request):
+    id = request.matchdict["id"]
```
Do not use `id` as variable name, `id` is builtin function. `rest_id` might be good candidate.

serge30 commented on easyrest/views/tag_controler.py:
```diff
@@ -0,0 +1,26 @@
+
+from sqlalchemy.exc import DBAPIError
+
+from json import dumps
```
PEP8: imports

serge30 commented on easyrest/views/tag_controler.py:
```diff
@@ -0,0 +1,26 @@
+    try:
+        query = request.dbsession.query(Tag)
+        tags = query.all()
+        if len(tags) == 0:
```
`if not tags:`

serge30 reviewed (changes requested):
Please install [pep8](https://pypi.org/project/pep8/) utility on your dev environment and check sources before commit. Also your may configure your IDE/Editor by installing corresponding plugins to check PEP8 compliance in the editor.

MaxSukhovetskyi committed:
fixed error with insertind id in error message

MaxSukhovetskyi committed:
Refactor tag_controler, make simpler

MaxSukhovetskyi committed:
Added docstring for tag_controller

MaxSukhovetskyi committed:
Changed models modules names acording to pep8

MaxSukhovetskyi committed:
Change len(tags) to not tags

MaxSukhovetskyi committed:
Added module docstring for config init file

MaxSukhovetskyi committed:
Stylies comments acording to PEP8

MaxSukhovetskyi committed:
Polished doctstring for models.meta module

MaxSukhovetskyi committed:
fix no such function error in tag_controler

MaxSukhovetskyi committed:
refactor return_all() function, added error if no restaurants in db

MaxSukhovetskyi committed:
fix typo error

MaxSukhovetskyi committed:
Change filter to .get(id), str concat to str interpolation

MaxSukhovetskyi committed:
Remove unnecessary function return all

MaxSukhovetskyi committed:
Json 404 page

MaxSukhovetskyi committed:
Changed json renderer to pyramid default, removed output id edit for React frontend

MaxSukhovetskyi committed:
Readme.md more detailed db init

MaxSukhovetskyi commented:
@serge30 I`ve finish all fixes. Make review please.

serge30 commented:
@MaxSukhovetskyi You still have a conflict. Please resolve it.

MaxSukhovetskyi committed:
Merge branch 'develop' into db_init_tags_model

MaxSukhovetskyi commented:
@serge30 We started to use new password it was a conflict. I`ve resolve it

MaxSukhovetskyi committed:
Init_db fix for uncorect string interpolation

MaxSukhovetskyi committed:
Modified readme.md init db for windows

MaxSukhovetskyi commented:
@serge30 Finaly ready for review)

serge30 commented on easyrest/views/default.py:
```diff
@@ -1,3 +1,7 @@
+"""Default controler for / route. Created by cookiecutter.
```
I do not think you still need it. Backend will not serve any pages, only reply on REST requests. You may remove this view and corresponding route.

serge30 commented on easyrest/views/restaurant_controler.py:
```diff
@@ -0,0 +1,126 @@
+    rests = request.dbsession.query(Restaurant).all()
+    rests_dict = asign_tags(rests)
+    if not rests_dict:
+        #body = wrap([], False, "No restaurants in database")
```
Please do not comment code, just remove it.

serge30 commented on easyrest/models/menu.py:
```diff
@@ -0,0 +1,26 @@
+    Defines data structure of "menu" table
+    Relationship: menu -> menu_item
+    """
+    __tablename__ = 'menu'
```
All table names should be in plural form: e.g. "menus". Model class name is correct here and should be in singular. Please update all models in this PR.

serge30 commented on easyrest/views/restaurant_controler.py:
```diff
@@ -0,0 +1,126 @@
+                    [<restaurant key:value>, ]
+                    "tags": [<tag dictionary>, ]
+                },
+            ]"""
```
PEP257: Closing quote (`"""`) should be on separate line.

https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings
> Unless the entire docstring fits on a line, place the closing quotes on a line by themselves

MaxSukhovetskyi committed:
Cleanups for pull request
Added raise HTTPNotFound to menu_controler,
Removed / view ad route,
Renamed db tables to plural form,
Fixed init_db crush that drop database do not drop relaited constrains
  (like Foreign keys),
Impruved docstrings.
